src/Concordium/Client/GRPC2.hs

```diff
@@ -0,0 +1,2766 @@
+            then Nothing
+            else return $ makeElectionDifficultyUnchecked af
+
+instance FromProto Proto.LeadershipElectionNonce where
+    type Output' Proto.LeadershipElectionNonce = LeadershipElectionNonce
+    fromProto = deMkSerialize
+
+instance FromProto Proto.ElectionInfo where
+    type Output' Proto.ElectionInfo = QueryTypes.BlockBirkParameters
+    fromProto eInfo = do
+        bbpBakers <-
+            mapM fromProto
+                . Vec.fromList
+                =<< eInfo ^? ProtoFields.bakerElectionInfo
+        bbpElectionDifficulty <- fromProto =<< eInfo ^? ProtoFields.electionDifficulty
+        bbpElectionNonce <- fromProto =<< eInfo ^? ProtoFields.electionNonce
+        return QueryTypes.BlockBirkParameters{..}
+
+instance FromProto Proto.NextUpdateSequenceNumbers where
+    type Output' Proto.NextUpdateSequenceNumbers = QueryTypes.NextUpdateSequenceNumbers
+    fromProto nums = do
+        _nusnRootKeys <- fromProto =<< nums ^? ProtoFields.rootKeys
+        _nusnLevel1Keys <- fromProto =<< nums ^? ProtoFields.level1Keys
+        _nusnLevel2Keys <- fromProto =<< nums ^? ProtoFields.level2Keys
+        _nusnProtocol <- fromProto =<< nums ^? ProtoFields.protocol
+        _nusnElectionDifficulty <- fromProto =<< nums ^? ProtoFields.electionDifficulty
+        _nusnEuroPerEnergy <- fromProto =<< nums ^? ProtoFields.euroPerEnergy
+        _nusnMicroCCDPerEuro <- fromProto =<< nums ^? ProtoFields.microCcdPerEuro
+        _nusnFoundationAccount <- fromProto =<< nums ^? ProtoFields.foundationAccount
+        _nusnMintDistribution <- fromProto =<< nums ^? ProtoFields.mintDistribution
+        _nusnTransactionFeeDistribution <- fromProto =<< nums ^? ProtoFields.transactionFeeDistribution
+        _nusnGASRewards <- fromProto =<< nums ^? ProtoFields.gasRewards
+        _nusnPoolParameters <- fromProto =<< nums ^? ProtoFields.poolParameters
+        _nusnAddAnonymityRevoker <- fromProto =<< nums ^? ProtoFields.addAnonymityRevoker
+        _nusnAddIdentityProvider <- fromProto =<< nums ^? ProtoFields.addIdentityProvider
+        _nusnCooldownParameters <- fromProto =<< nums ^? ProtoFields.cooldownParameters
+        _nusnTimeParameters <- fromProto =<< nums ^? ProtoFields.timeParameters
+        return QueryTypes.NextUpdateSequenceNumbers{..}
+
+-- An IP address.
+type IpAddress = Text
+
+-- An IP port.
+type IpPort = Int16
+
+-- A peer are represented by its IP address.
+type Peer = IpAddress
+
+-- Compound type representing a pair of an IP address and a port.
+type IpSocketAddress = (IpAddress, IpPort)
+
+instance ToProto IpAddress where
+    type Output IpAddress = Proto.IpAddress
+    toProto ip = Proto.make $ ProtoFields.value .= ip
+
+instance ToProto IpPort where
+    type Output IpPort = Proto.Port
+    toProto ip = Proto.make $ ProtoFields.value .= fromIntegral ip
+
+instance FromProto Proto.IpAddress where
+    type Output' Proto.IpAddress = IpAddress
+
+    -- FIXME: Should this be validated?
```

Comment:
Not now. this is just going to be printed.

If we did we should get find a library (does prelude define it, or network?) that defines an IP data type. We are definitely not writing our own IP address validation.

Reply:
Anyway, just ignore validation here since you are mapping it to a string anyhow.